### PR TITLE
Feature/114/tables

### PR DESCRIPTION
**Pull request by max-nextcloud**
* Resolves: #114
* Target version: master 
Bildschirmfoto von 2022-03-22 11-24-29

### CODE REVIEW

File: src/nodes/Table.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		if (nodeType.spec.tableRole) {
+			roles[nodeType.spec.tableRole] = nodeType
+		}
+	})
+
+	schema.cached.tableNodeTypes = roles
+
+	return roles
+}
+
```
Comment (max-nextcloud) [inline src/nodes/Table.js:52]: Asked for exporting this in https://github.com/ueberdosis/tiptap/issues/2621

**Comment by juliushaertl** [comment]
Quick early design feedback form the spontaneous design co-working:

- for the header, only the bottom border should be darker, possibly no background color
- there should be a border radius on the table
- some margin at the bottom would help to separate from the next block
- keyboard actions
  - enter in the last row should create a new row
  - tab navigation should work within cells
- when the table is focussed, buttons could be shown for adding/removing rows/columns outside of the table
- https://materialdesignicons.com/icon/table might be a nicer icon

File: src/nodes/Table.js (hunk 1 above)
Comment (max-nextcloud) [reply src/nodes/Table.js:52]: Stopped using this as `tableRoles` are used multiple times in our implementation now. We need the row type for the first line that only includes `<th>` and the other rows as well. So identifying node types by table role is no longer a good idea.

**Comment by max-nextcloud** [comment]
Enter already creates a new row almost everywhere in the table.
It does not work in the last cell of a row though and not if the cell contains text.

- [ ] make it navigate to the next row everywhere but in the last row
- [ ] make it create a new row in the last row - even if the cell has content
- [ ] make it create a new row in the last row even in the last cell.

**Comment by max-nextcloud** [comment]
Okay... we're mostly missing the adding and removal of columns and rows with buttons now.

Adding them next to the table will require turning the respective table elements into [node views](https://tiptap.dev/guide/node-views/vue). This seems doable but also pretty tedious. Might go with a more simple version of just adding the coresponding buttons to a table drop down menu for NC 24.

**Comment by max-nextcloud** [comment]
What's missing to get this ready to merge:
- [ ] fix markdown unit tests
- [ ] fix cypress tests by telling cypress how to handle vue files (possibly turn this into a component test)
- [ ] add button to delete the table
- [ ] rebase & compile

Further tweaks that would be nice to have but could be fixed separately
- [ ] use `color-primary-light` for highlighting current row
- [ ] check if rows and columns can be added / removed and disable buttons accordingly
- [ ] refactor test layout, put all tables fixtures in one folder
- [ ] accessibility for adding columns and rows (currently tab moves past the buttons in the cells)

**Comment by max-nextcloud** [comment]
I think this is good (enough) to go. I'd be curious what you think.
I'll see if i can get to fix some of the open tasks left while addressing other things that come up during a review. Otherwise i'd move them to a new issue and fix them later after this has landed.

**Comment by max-nextcloud** [comment]
Looks like the test failures are not related to this branch... they also happen on master.

**Comment by mejo-** [comment]
Wow, looks super nice :star_struck:

I only gave it a quick try and it looks really great! Some glitches I experienced so far:

* Adding a table at the very top of the document, I'm no longer able to put something on top if it afterwards. Would be nice if I could put the cursor above the table somehow
* Navigating with the arrow up (or left) key from below the table, the cursor ends up on the right side of the table. I'm able to add content there. That's unexpected and I'd consider it unwanted ;)
* Further navigating up with the arrow up keys, the lower right field first is reached first and then it goes up the very right column until the last body field. I'm not able to jump into the header field by navigating with the cursors.
* If the table is the first node in my document and I try to navigate further up from a header field or the last body field, the cursor jumps below the table.

**Review by juliushaertl — COMMENTED** [review]
Only two minor comments from my side, will continue with some testing

**Comment by juliushaertl** [comment]
Something for a later iteration maybe would be to also support left/center/right alignment as described in https://github.github.com/gfm/#delimiter-row

File: src/components/EditorWrapper.vue [hunk 2]
```diff
@@ -575,7 +575,6 @@ export default {
 		&:not(.show-color-annotations)::v-deep .author-annotation,
 		&:not(.show-color-annotations)::v-deep .image {
 			background-color: transparent !important;
-			color: var(--color-main-text) !important;
```
[the remark this answers is not in the document]
Comment (max-nextcloud) [reply src/components/EditorWrapper.vue:578]: But this is just removing the `color` property... not the `background-color` of the rule. So i think it should not affect background colors. Or is image somehow using `color` for the background?

File: css/prosemirror.scss [hunk 3]
```diff
@@ -252,6 +252,57 @@ div.ProseMirror {
 		}
 	}
 
+	// TODO: create new css variables to improve the theming capability.
```
[the remark this answers is not in the document]
Comment (vinicius73) [reply css/prosemirror.scss]: Thanks for remembering @juliushaertl, I've already changed it.

**Comment by eneiluj** [comment]
Amazing!

A few successful checks:
* [x] Images are displayed correctly in cells!
* [x] Emoji completion works
* [x] It rocks

A few non-blocking remarks:
* As only a subset of the formatting features are possible in table cells, we could find a way to make it explicit for the user. It feels a little bit frustrating to have no feedback when trying to create a list in a cell. We could disable/hide the related menu buttons or show an error message.
* When adding a column or a row, the actions menu stays open. It feels weird :grin:
* When trying to select the content of multiple cells, 2 cells start having a green border and it quickly disappears.
* The bottom border of the header row could be bigger to make it even more explicit it's a header. Or we could make all the borders or the header thicker.

**Comment by max-nextcloud** [comment]
@eneiluj Would you approve this? I'll rebase and push again. Would be nice if it could still land in beta2.

**Review by mejo- — APPROVED** [review]

**Pull request merged**